### PR TITLE
feat: get cds overlap given chromosome, start, + stop

### PR DESCRIPTION
**Pull request by korikuzma**
Address #216 for 0.1.x . This is needed for variation manuscript

Notes:
* Add class (`FeatureOverlap`) for getting cds overlap
  * Only supports GRCh38 input
  * Returns VRS Sequence Locations 
* `get_mane_summary` --> `get_mane`
  * Now can additionally download MANE RefSeq GFF file

### CODE REVIEW

File: cool_seq_tool/data_sources/feature_overlap.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            `identifier`. If both `chromosome` and `identifier` are provided,
+            `chromosome` will be used.
+        :param identifier: Genomic identifier on GRCh38 assembly. If not provided,
+            must identifier `chromosome`. If both `chromosome` and `identifier` are
+            provided, `chromosome` will be used.
+        :param residue_mode: Residue mode for `start` and `end`
+        :raise FeatureOverlapError: If missing required fields
+        :return: Feature overlap dictionary where the key is the gene name and the value
+            is the list of CDS overlap (cds_start, cds_stop, overlap_start,
+            overlap_stop). Will return residue coordinates.
```
Comment (korikuzma) [inline cool_seq_tool/data_sources/feature_overlap.py:134]: @wesleygoar @austinant Wasn't sure if we wanted to return residue or inter-residue coordinates

File: cool_seq_tool/data_sources/feature_overlap.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                )
+
+        # GFF is 1-based, so we need to convert inter-residue to residue
+        # RESIDUE           |   | 1 |   | 2 |   | 3 |   |
+        # INTER_RESIDUE     | 0 |   | 1 |   | 2 |   | 3 |
+        if residue_mode == ResidueMode.INTER_RESIDUE:
+            if start != end:
+                start += 1
+            else:
+                end += 1
```
Comment (korikuzma) [inline cool_seq_tool/data_sources/feature_overlap.py:154]: I'm not sure about `start == end`. I was going off of HGVS insertion example: http://normalize.cancervariants.org/variation/translate_from?variation=ENST00000331728.9:c.2049_2050insA

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 1 above)
Comment (wesleygoar) [reply cool_seq_tool/data_sources/feature_overlap.py:134]: Hmm good question. We do inter-residue for most of our other operations right? @ahwagner do you want to weight in on this?

Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py:134]: Yes. I'll wait until @ahwagner replies

Comment (wesleygoar) [reply cool_seq_tool/data_sources/feature_overlap.py:134]: @austinant did you do this initially with residue based coordinates?

Comment (wesleygoar) [reply cool_seq_tool/data_sources/feature_overlap.py:134]: @korikuzma sounds good. Pinging @ahwagner again just to make sure he sees this haha

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 2 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py:154]: @ahwagner Can you help with this too? 😅

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 1 above)
Comment (ahwagner) [reply cool_seq_tool/data_sources/feature_overlap.py:134]: Let's use VRS Sequence Location objects for describing regions (so, interresidue). I don't know where SA is at (I think they just created a JSON schema PR today–haven't reviewed) but I know they will be using VRS Sequence Locations to describe features.

Comment (austinant) [reply cool_seq_tool/data_sources/feature_overlap.py:134]: @wesleygoar i used whatever coords are output by normalizer (for the variant) + whatever coords in the GFF file (for the CDS region)

File: cool_seq_tool/data_sources/feature_overlap.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    def get_grch38_cds_overlap(
+        self,
+        start: int,
+        end: int,
+        chromosome: Optional[str] = None,
+        identifier: Optional[str] = None,
+        residue_mode: ResidueMode = ResidueMode.RESIDUE,
+    ) -> Optional[Dict]:
+        """Get feature overlap for GRCh38 genomic data
```
Comment (jsstevenson) [inline cool_seq_tool/data_sources/feature_overlap.py:120]: Would someone (Kori? @austinant?) be willing -- for my sake -- to add a descriptive sentence or two here? I can also paraphrase something from the manuscript if there's a specific relevant section

File: cool_seq_tool/data_sources/feature_overlap.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+        :param start: GRCh38 start position
+        :param end: GRCh38 end position
+        :param chromosome: Chromosome. 1..22, X, or Y. If not provided, must provide
+            `identifier`. If both `chromosome` and `identifier` are provided,
+            `chromosome` will be used.
+        :param identifier: Genomic identifier on GRCh38 assembly. If not provided,
+            must identifier `chromosome`. If both `chromosome` and `identifier` are
```
Comment (jsstevenson) [inline cool_seq_tool/data_sources/feature_overlap.py:128]: Is there a word missing in "If not provided, must identifier `chromosome`"?

Comment (jsstevenson) [reply cool_seq_tool/data_sources/feature_overlap.py:128]: Genome needs identified @austinant

Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py:128]: Nice catch. I accidentally replaced the wrong word. If not provided, must provide chromosome

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 2 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py:154]: @ahwagner yes

Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py:154]: Well, maybe not for this method (tbd). But I’d like still like to know the answer

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 3 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: @jsstevenson updated in latest commit. Let me know if that clears things up for you

**Comment by korikuzma** [comment]
Need to look over this with fresh brain/eyes before re-requesting reviews

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 2 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: I ended up modifying the GFF dataframe by substracting one for the start positions: 
https://github.com/GenomicMedLab/cool-seq-tool/blob/7e4272828d6f1bc962847a6b3aa22a5d645f4385/cool_seq_tool/data_sources/feature_overlap.py#L75-L76

Then modifying user's residue input:
https://github.com/GenomicMedLab/cool-seq-tool/blob/7e4272828d6f1bc962847a6b3aa22a5d645f4385/cool_seq_tool/data_sources/feature_overlap.py#L159-L164

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 1 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: @austinant I can't remember if you had accounted for GFF using 1 based and Variation Normalizer output using inter-residue when getting the overlap, but I addressed this in my latest changes.

Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: Updated so that VRS Sequence Locations are used in the response.

**Comment by korikuzma** [comment]
> Need to look over this with fresh brain/eyes before re-requesting reviews

Ready for review!

**Review by wesleygoar — APPROVED** [review]

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 2 above)
Comment (ahwagner) [reply cool_seq_tool/data_sources/feature_overlap.py]: Not sure about lines 161-162. How does one specify a 1nt feature?

File: cool_seq_tool/data_sources/feature_overlap.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                    ga4gh_seq_id = identifier
+            else:
+                raise FeatureOverlapError(
+                    "Must provide either `chromosome` or `identifier`"
+                )
+
+        # Convert residue to inter-residue
+        if residue_mode == ResidueMode.RESIDUE:
+            if start == end:
+                end -= 1
```
Comment (ahwagner) [inline cool_seq_tool/data_sources/feature_overlap.py:166]: If there is a 1 nt search, how is that specified?

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 2 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: @ahwagner Idk I was just going with if someone passed start=1 and end=1. Is that not something that could happen? If not, I can raise an exception if start == stop.

Comment (ahwagner) [reply cool_seq_tool/data_sources/feature_overlap.py]: I am just wondering why we handle the case of `start == end` explicitly. In residue coordinates, this represents a 1nt search, which is totally possible. But if we then do `end -= 1`, it is now a 0 nt search, which may not be what was intended. So trying to understand the requirements for the `start == end` block.

Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: I was thinking of 0-based, which needs to be fixed in a separate issue.

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 5 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: @ahwagner pushed fix. let me know if this is what you were expecting

**Review by ahwagner — APPROVED** [review]

**Comment by wesleygoar** [comment]
Good catch @ahwagner

**Comment by korikuzma** [comment]
> Good catch @ahwagner

We have to keep @ahwagner  on his toes

**Review by austinant — APPROVED** [review]

File: cool_seq_tool/data_sources/feature_overlap.py (hunk 3 above)
Comment (korikuzma) [reply cool_seq_tool/data_sources/feature_overlap.py]: @austinant how about

```python
        """Given GRCh38 genomic data, find the overlapping MANE features (gene and cds).
        The genomic data is specified as a sequence location by `chromosome`, `start`,
        `end`. All CDS regions with which the input sequence location has nonzero base
        pair overlap will be returned.

        :return: MANE feature (gene/cds) overlap data represented as a dict. The
            dictionary will be keyed by genes which overlap the input sequence location.
            Each gene contains a list of the overlapping CDS regions with the beginning
            and end of the input sequence location's overlap with each
            {
                gene: {
                    'cds': VRS Sequence Location
                    'overlap': VRS Sequence Location
                }
            }
        """
```

**Pull request merged**